File: 10_flutter_and_native_samples/android/app/src/main/kotlin/com/example/flutter_and_native_samples/NativeTickerChannel.kt
package com.example.flutter_and_native_samples

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

/**
 * Pushes a tick event to Flutter every second while a listener is active.
 *
 * Demonstrates EventChannel lifecycle: [onListen] starts the timer,
 * [onCancel] stops it.
 */
class NativeTickerChannel(
    messenger: BinaryMessenger,
) : EventChannel.StreamHandler {
    private val channel =
        EventChannel(messenger, "examples.flutter_native_calls/ticker")

    private val mainHandler = Handler(Looper.getMainLooper())
    private var events: EventChannel.EventSink? = null
    private var tick = 0

    private val tickerRunnable =
        object : Runnable {
            override fun run() {
                val sink = events ?: return
                tick += 1
                sink.success(
                    mapOf(
                        "tick" to tick,
                        "timestampMs" to System.currentTimeMillis(),
                    ),
                )
                mainHandler.postDelayed(this, 1_000L)
            }
        }

    fun register() {
        channel.setStreamHandler(this)
    }

    fun unregister() {
        stopTicker()
        channel.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.events = events
        tick = 0
        // Emit the first tick immediately, then every second.
        mainHandler.post(tickerRunnable)
    }

    override fun onCancel(arguments: Any?) {
        stopTicker()
    }

    private fun stopTicker() {
        mainHandler.removeCallbacks(tickerRunnable)
        events = null
        tick = 0
    }
}

File: 10_flutter_and_native_samples/ios/Runner/NativeTickerChannel.swift
import Flutter
import Foundation

/// Pushes a tick event to Flutter every second while a listener is active.
///
/// Demonstrates EventChannel lifecycle: `onListen` starts the timer,
/// `onCancel` stops it.
final class NativeTickerChannel: NSObject, FlutterStreamHandler {
  static let channelName = "examples.flutter_native_calls/ticker"

  private let channel: FlutterEventChannel
  private var events: FlutterEventSink?
  private var timer: Timer?
  private var tick = 0

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterEventChannel(name: Self.channelName, binaryMessenger: messenger)
    super.init()
  }

  func register() {
    channel.setStreamHandler(self)
  }

  func unregister() {
    stopTicker()
    channel.setStreamHandler(nil)
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    self.events = events
    tick = 0
    // Emit the first tick immediately, then every second on the main run loop.
    let timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
      self?.emitTick()
    }
    self.timer = timer
    emitTick()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    stopTicker()
    return nil
  }

  private func emitTick() {
    guard let events else {
      return
    }
    tick += 1
    let timestampMs = Int(Date().timeIntervalSince1970 * 1_000)
    events([
      "tick": tick,
      "timestampMs": timestampMs,
    ])
  }

  private func stopTicker() {
    timer?.invalidate()
    timer = nil
    events = nil
    tick = 0
  }
}
